dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/internal/AppTrackerImpl.java

```diff
@@ -238,6 +253,63 @@ public boolean isStarted(String appName) {
         }
     }
 
+    /**
+     * Returns true if the application with the specified name is installed, otherwise false.
+     *
+     * @return true if the application with the specified name is installed, otherwise false.
+     */
+    @Override
+    public boolean isInstalled(String appName) {
+        lock.readLock().lock();
+        try {
+            if (appStateMap.get(appName) == ApplicationState.INSTALLED) {
+                String state = getApplicationMBean(appName);
```

Comment:
Do we need to call the MBean here, since we are calling it in the applicationStopped() method? I remember discussing about it, can you refresh my memory, why we need it here, instead of just reading the appStateMap similar to the isStarted() method. We call this method in the HealthCheckServiceImpl, so if the endpoint is hit every 5 seconds, the MBean would get called every 5 seconds, which might cause a performance impact.

Reply:
This should only have a performance impact for INSTALLED apps. Since INSTALLED apps are essentially apps that failed to start, they have already called applicationStopped() but are not stopped yet. (In general, I am assuming most apps should be STARTING (delayed) or STARTED (running), and for those cases, the MBean would not get checked, so no performance impact) The apps check by this if statement are in a "leaked" state where we won't know if/when they shutdown, so we need to check the MBean to see if they are removed.

Reply:
That makes sense, thanks!